src/privileged/vpnRecommender/api.js

```diff
@@ -0,0 +1,562 @@
+    }
+
+    if (message.name === "VpnRecommender::dismiss") {
+      eventName = "dismiss";
+      this.concludeNotification("dismiss");
+    }
+
+    if (message.name === "VpnRecommender::info") {
+      eventName = "info";
+    }
+
+    if (message.name === "VpnRecommender::autoDismiss") {
+      eventName = "auto-dismiss";
+      this.concludeNotification("auto-dismiss");
+      this.sendTelemetry({
+        "message_type": "auto_dismissal_event",
+        "reason": message.data.reason,
+      });
+    }
+
+    if (message.name === "VpnRecommender::dontShowChange") {
+      eventName = "dont-show-change";
+      this.dontShowChange(message.data.checked);
+    }
+
+    this.telemetryCallback({
+      "message_type": "event",
+      "event": eventName,
+    });
+  }
+
+  getTriggerData(trigger, pref) {
+    const str_val = Preferences.get(pref);
+    if (!str_val) { // pref does not exist
+      return str_val;
+    }
+
+    const val = JSON.parse(str_val);
+    return val[trigger];
+  }
+
+  setTriggerData(trigger, pref, value) {
+    let str_val = Preferences.get(pref);
+    if (!str_val) {
+      str_val = "{}";
+    }
+
+    const obj_val = JSON.parse(str_val);
+    obj_val[trigger] = value;
+    Preferences.set(pref, JSON.stringify(obj_val));
+  }
+
+  tryShowNotification(trigger) {
+    const isShadow = (trigger !== this.variation);
+
+    this.sendTelemetry({
+      "message_type": "event",
+      "event": "trigger",
+      "trigger": trigger,
+      "is-shadow": String(isShadow),
+    });
+
+    if (Date.now() - Number(this.getTriggerData(trigger, LAST_NOTIFICATION_PREF)) < TWENTY_FOUR_HOURS) {
+      log("less than 24 hours has passed since the last notification was shown");
+      return;
+    }
+
+    const notificationCount = this.getTriggerData(trigger, NOTIFICATION_COUNT_PREF) || 0;
+    if (notificationCount === MAX_NOTIFICATION_COUNT) return;
+
+    if (!this.RecentWindow.getMostRecentBrowserWindow()) {
+      return; // if all windows are private
+    }
+
+    // this condition ensures that the doorhanger is actually shown (not a shadow notification) only when the trigger and the branch match
+    if (!isShadow && !Preferences.get(DONT_SHOW_PREF)) {
+      this.showNotification();
+
+      this.notificationResultData = {
+        dontShowChecked: false,
+        notificationNumber: notificationCount + 1,
+        result: "unknown",
+      };
+
+      this.sendTelemetry({
+        "message_type": "event",
+        "event": "notification-delivered",
+        "number": String(notificationCount + 1),
+      });
+    }
+
+    this.setTriggerData(trigger, NOTIFICATION_COUNT_PREF, notificationCount + 1); // increment notification count
+    this.setTriggerData(trigger, LAST_NOTIFICATION_PREF, String(Date.now()));
+
+    this.sendTelemetry({
+      "message_type": "event",
+      "event": "shadow-notification",
+      "number": String(notificationCount + 1),
+      "trigger": trigger,
+      "is-shadow": String(isShadow),
+    });
```

Comment:
I guess it's mostly up to you, but why are you sending this notification in all cases? The data in here should be covered by the previous events in this function in the no-shadow case.